File: src/main/java/leetcode/lists/hot100/LeetCode2.java
package leetcode.lists.hot100;

import common.*;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 23:11
 */
public class LeetCode2 {

    // fix case
    //[9,9,9,9,9,9,9]
    //[9,9,9,9]


    int flag = 0;

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        addTwoNumbers(l1, l2, dummy);
        return dummy.next;
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2, ListNode dummy) {
        if (l1 == null && l2 == null) {
            if (flag == 1) {
                dummy.next = new ListNode(1);
                return dummy;
            }
        }
        int num1 = l1 == null ? 0 : l1.val;
        int num2 = l2 == null ? 0 : l2.val;

        int sum = num1 + num2 + flag;
        flag = sum / 10;

        ListNode node = new ListNode(sum % 10);
        dummy.next = node;
        return addTwoNumbers(l1 == null ? null : l1.next, l2 == null ? null : l2.next, node);
    }
}

File: go/src/leetcode/hot100/LeetCode2.go
package hot100

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	root := &ListNode{}
	helperAdd(l1, l2, root, 0)
	return root.Next
}

func helperAdd(l1 *ListNode, l2 *ListNode, root *ListNode, flag int) {
	if l1 == nil && l2 == nil {
		if flag == 1 {
			root.Next = &ListNode{Val: 1}
		}
		return
	}
	var v1, v2 int
	var n1, n2 *ListNode
	if l1 == nil {
		v1 = 0
		n1 = nil
	} else {
		v1 = l1.Val
		n1 = l1.Next
	}
	if l2 == nil {
		v2 = 0
		n2 = nil
	} else {
		v2 = l2.Val
		n2 = l2.Next
	}
	root.Next = &ListNode{
		Val: (v1 + v2 + flag) % 10,
	}
	flag = (v1 + v2 + flag) / 10
	helperAdd(n1, n2, root.Next, flag)
}
